monoio-http/src/common/parsed/request.rs

```diff
@@ -157,25 +157,46 @@ impl<P> ParsedRequest<P> {
             *url_params = Parse::Parsed(Default::default());
             return Ok(());
         };
-        let parsed = serde_urlencoded::from_str(query).map_err(|e| {
+
+        let map: Vec<(String, String)> = serde_urlencoded::from_str(query).map_err(|e| {
             *url_params = Parse::Failed;
             e
         })?;
-        *url_params = Parse::Parsed(parsed);
+
+        let mut qmap = QueryMap::new();
+        for (key, value) in map {
+            qmap.entry(key).or_insert(Vec::new()).push(value);
+        }
+        *url_params = Parse::Parsed(qmap);
         Ok(())
     }
 
     /// Try parse if not parsed before, then get the value of the URL parameter.
     #[inline]
-    pub fn get_url_param(&self, name: &str) -> Result<Option<String>, ParseError> {
+    pub fn get_url_param(&self, name: &str) -> Result<Option<&str>, ParseError> {
+        self.parse_url_params()?;
+        unsafe {
+            let parsed = &*self.url_params.get();
+            Ok(parsed
+                .as_ref()
+                .unwrap_unchecked()
+                .get(name)
+                .and_then(|v| v.iter().next().map(|s| s.as_str())))
+        }
+    }
+
+    pub fn get_url_params_all<'a>(
+        &'a self,
+        name: &str,
+    ) -> Result<Option<impl Iterator<Item = &'a str>>, ParseError> {
```

Comment:
`impl Iterator` means users cannot random access it. Maybe we can expose `&'a [String]`?
Yes it is indeed inconsistent with the interface for obtaining a single value. But it can expose more capabilities.

Reply:
Changed params_all to return &[String]